apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,225 @@
+
+import { Modal } from '../../src/components/Modal.html.js';
+
+import { AccessRulesTable } from './lib/accessRulesTable.js';
+import { EditAccessRuleModal } from './lib/editAccessRuleModal.js';
+
+onDocumentReady(() => {
+  const enableEditButton = document.getElementById('enableEditButton');
+  const editModeButtons = document.getElementById('editModeButtons');
+  const accessRulesTable = document.querySelector('.js-access-rules-table');
+  const editAccessRuleModal = document.querySelector('.js-edit-access-rule-modal');
+  const deleteAccessRuleModal = document.querySelector('.js-delete-access-rule-modal');
+
+  const accessRulesData = decodeData('access-rules-data');
+
+  const ptHost = (accessRulesTable as HTMLElement)?.dataset.ptHost ?? '';
+  const devMode = (accessRulesTable as HTMLElement)?.dataset.devMode === 'true';
+  const hasCourseInstancePermissionView =
+    (accessRulesTable as HTMLElement)?.dataset.hasCourseInstancePermissionView === 'true';
+  const timezone = (accessRulesTable as HTMLElement)?.dataset.timezone ?? 'UTC';
+
+  let editMode = false;
+
+  function refreshTable() {
+    morphdom(
+      accessRulesTable as Node,
+      AccessRulesTable({
+        accessRules: accessRulesData,
+        ptHost,
+        devMode,
+        hasCourseInstancePermissionView,
+        editMode,
+      }).toString(),
+    );
+  }
+
+  refreshTable();
+
+  const addRuleButton = document.getElementById('addRuleButton');
+
+  enableEditButton?.addEventListener('click', () => {
+    editMode = true;
+    enableEditButton.style.display = 'none';
+    editModeButtons?.style.removeProperty('display');
+    addRuleButton?.style.removeProperty('display');
+    refreshTable();
+  });
+
+  on('click', '#saveAndSyncButton', () => {
+    const form = document.getElementById('accessRulesForm') as HTMLFormElement;
+    const assessmentAccessRulesInput = form.querySelector(
+      'input[name="assessment_access_rules"]',
+    ) as HTMLInputElement;
+    if (!assessmentAccessRulesInput) return;
+    accessRulesData.forEach((accessRule: Record<string, any>) => {
+      accessRule.startDate = adjustedDate(accessRule.start_date).toISOString().slice(0, 19);
+      accessRule.endDate = adjustedDate(accessRule.end_date).toISOString().slice(0, 19);
+      accessRule.timeLimitMin = parseInt(accessRule.time_limit);
+      accessRule.uids = accessRule.uids.split(',').map((uid: string) => uid.trim());
+      accessRule.credit = parseInt(accessRule.credit);
+      if (accessRule.exam) {
+        accessRule.examUuid = accessRule.exam_uuid;
+      }
+      delete accessRule.formatted_start_date;
+      delete accessRule.formatted_end_date;
+      delete accessRule.start_date;
+      delete accessRule.end_date;
+      delete accessRule.time_limit;
+      delete accessRule.pt_course_name;
+      delete accessRule.pt_exam_name;
+      delete accessRule.pt_course_id;
+      delete accessRule.pt_exam_id;
+      delete accessRule.exam_uuid;
```

Comment:
Instead of implementing all this as `forEach` with mutation in each iteration, consider instead implementing it as a `map()` where you return the specific properties you want for each serialized access rule.

Reply:
Updated to a `map()` in [1dfef07](https://github.com/PrairieLearn/PrairieLearn/pull/9934/commits/1dfef07527ed0296ff3cdfa7c58d85fce2304b97) to serialize each rule.